### ISSUE TITLE
Share some connection infrastructure with Fetch

### ISSUE THREAD

**Issue by annevk** [w3c/webrtc-extensions#73]
I think there are basically two reasons why we would want centralized connection management in the platform:

* Consistent handling of overrides of resolving names (domains), e.g., mapping localhost to the loopback address, as discussed at https://github.com/whatwg/fetch/issues/1118.
* Consistent requirements on partitioning of all aspects of the connection. https://fetch.spec.whatwg.org/#network-partition-keys in Fetch is now used for connections and the HTTP cache, but it should also apply to the DNS cache, etc. Having those kind of requirements once would be beneficial. Also in case we ever decide to adjust them.

(See also https://github.com/w3c/webtransport/issues/128.)

**Comment by alvestrand** [comment]
The network partition key (https://fetch.spec.whatwg.org/#network-partition-key) may be a sharable object.

**Comment by jan-ivar** [comment]
@annevk to unstuck this, would you be able to provide examples or guidance (or even a PR) with how and which algorithms you imagine WebRTC invoking in fetch here?

**Comment by annevk** [comment]
I'm slowly making progress on enabling this together with @ricea. The idea is that WebRTC would at some point hook into https://fetch.spec.whatwg.org/#concept-connection-obtain (or perhaps something lower-level as mt points out above), with parameters that make sense for it. That would then take care of any domain resolving (if needed) and setting up and returning a connection.

My hope is that we can also move some other logic there, such as port blocking, what `localhost` means, etc. (I think since mt's comment above implementations have reached agreement on sharing port blocking logic across HTTP/WebRTC/WebTransport connections, with the exception of one or two ports that are safelisted for WebRTC use.)

**Comment by alvestrand** [comment]
Anne, I'm having a bit of trouble understanding how "obtain a connection" could be a common merge point.
To start off with, WebRTC doesn't use hostnames in the common use case (it does use MDNS hostnames in some cases, and it's possible to use hostnames too, but the by-far majority of usages use IP addresses).

I do appreciate that @ricea is trying to help, but having someone who knows the WebRTC networking stack (ICE, STUN and all that machinery) in the loop is probably vital, so that we don't generate proposals that don't make sense in the WebRTC context.

**Comment by annevk** [comment]
Agreed. We're still in the process of ensuring Fetch, WebSocket, and WebTransport are aligned and it might make sense that WebRTC only reuses some aspects. I'd love to chat once we're closer. Here are the things that come to mind with regards to sharing:

* Keying. Any kind of cache and connection pooling should take into account the network partition key.
* Port blocking. While not currently the case, it's our plan to enforce this while obtaining a connection.
* Proxies. It would be ideal that if there's a proxy and it doesn't do UDP, the browser doesn't attempt to do UDP. (To be clear, this could use input. It's not clear how realistic this is.)
* CSP. I think it would make sense if part of CSP was enforced here.
* online/offline events. These events are somewhat problematic as they fire around the same time across tabs, creating a cross-site communication channel and privacy leak. It would be better if they only fired after the global state change once a tab is focused by the user. However, unless we also make and enforce the state per tab, this would be meaningless. (It's still somewhat tricky to avoid leaks entirely in the going offline case, but I think we can improve on the status quo at least.)

**Comment by zenhack** [comment]
@annevk, is your WIP online in some kind of draft document?

**Comment by zenhack** [comment]
Ping?

**Comment by annevk** [comment]
Status is still the same as I've been OOO. You can check connection PRs at https://github.com/whatwg/fetch/pulls for progress with regards to the first paragraph.
